packages/core/source/legacy/host.ts

```diff
@@ -0,0 +1,335 @@
+      const [mutationType, parentId] = record;
+
+      switch (mutationType) {
+        case MUTATION_TYPE_INSERT_CHILD: {
+          const node = record[2];
+          const index = record[3];
+          persistNode(parentId, node, index);
+          break;
+        }
+        case MUTATION_TYPE_REMOVE_CHILD: {
+          const index = record[2];
+          removeNode(parentId, index);
+          break;
+        }
+      }
+    }
+
+    connection.mutate(records);
+  }
+
+  return function remoteChannel<T extends keyof LegacyActionArgumentMap>(
+    type: T,
+    ...payload: LegacyActionArgumentMap[T]
+  ) {
+    switch (type) {
+      case LEGACY_ACTION_MOUNT:
+        const [nodes] =
+          payload as LegacyActionArgumentMap[typeof LEGACY_ACTION_MOUNT];
+
+        const records = nodes.map(
+          (node, index) =>
+            [
+              MUTATION_TYPE_INSERT_CHILD,
+              ROOT_ID,
+              adaptLegacyNodeSerialization(node, options),
+              index,
+            ] satisfies RemoteMutationRecord,
+        );
+
+        mutate(records);
+
+        break;
+
+      case LEGACY_ACTION_INSERT_CHILD:
+        const [parentId, index, child] =
+          payload as LegacyActionArgumentMap[typeof LEGACY_ACTION_INSERT_CHILD];
+
+        mutate([
+          [
+            MUTATION_TYPE_INSERT_CHILD,
+            parentId ?? ROOT_ID,
+            adaptLegacyNodeSerialization(child, options),
+            index,
+          ],
+        ]);
+
+        break;
+
+      case LEGACY_ACTION_REMOVE_CHILD: {
+        const [parentID, removeIndex] =
+          payload as LegacyActionArgumentMap[typeof LEGACY_ACTION_REMOVE_CHILD];
+
+        mutate([
+          [MUTATION_TYPE_REMOVE_CHILD, parentID ?? ROOT_ID, removeIndex],
+        ]);
+
+        break;
+      }
+
+      case LEGACY_ACTION_UPDATE_TEXT: {
+        const [textId, text] =
+          payload as LegacyActionArgumentMap[typeof LEGACY_ACTION_UPDATE_TEXT];
+
+        mutate([[MUTATION_TYPE_UPDATE_TEXT, textId, text]]);
+
+        break;
+      }
+
+      case LEGACY_ACTION_UPDATE_PROPS: {
+        const [id, props] =
+          payload as LegacyActionArgumentMap[typeof LEGACY_ACTION_UPDATE_PROPS];
+
+        const records = [];
+
+        for (const [key, value] of Object.entries(props)) {
+          if (isFragment(value)) {
+            const index = tree[id]?.findIndex(({slot}) => slot === key) ?? -1;
+
+            if (index !== -1) {
+              records.push([
+                MUTATION_TYPE_REMOVE_CHILD,
+                id,
+                index,
+              ] satisfies RemoteMutationRecord);
+            }
+
+            records.push([
+              MUTATION_TYPE_INSERT_CHILD,
+              id,
+              adaptLegacyFragmentSerialization(key, value, options),
+              tree[id]?.length ?? 0,
+            ] satisfies RemoteMutationRecord);
+          } else {
+            const index = tree[id]?.findIndex(({slot}) => slot === key) ?? -1;
```

Comment:
Should we lift this out since we have to deal with it in all cases? This also makes me wonder if we would be better off with a map of slot => index, so that we wouldn't need to loop through every child on each property update.

Reply:
If the map is reduced to slot => index then we can't do a clean up when some slot is removed as we can't know if some other slot was a nested child of the removed slot.